openml/extensions/sklearn/extension.py

```diff
@@ -1646,20 +1673,28 @@ def _prediction_to_probabilities(y: np.ndarray, classes: List[Any]) -> np.ndarra
                     # then we need to add a column full of zeros into the probabilities
                     # for class 3 because the rest of the library expects that the
                     # probabilities are ordered the same way as the classes are ordered).
-                    proba_y_new = np.zeros((proba_y.shape[0], len(task.class_labels)))
-                    for idx, model_class in enumerate(model_classes):
-                        proba_y_new[:, model_class] = proba_y[:, idx]
-                    proba_y = proba_y_new
-
-                if proba_y.shape[1] != len(task.class_labels):
                     message = "Estimator only predicted for {}/{} classes!".format(
                         proba_y.shape[1], len(task.class_labels),
                     )
                     warnings.warn(message)
                     openml.config.logger.warn(message)
+
+                    # True if predict_proba is successfully called with X_test as numpy
+                    # if X_test is dataframe, proba_y will be dataframe with model_classes columns
+                    if isinstance(proba_y, np.ndarray):
+                        proba_y = pd.DataFrame(proba_y, columns=model_classes)
+
+                    for i, col in enumerate(task.class_labels):
+                        # adding missing columns with 0 probability
+                        if col not in model_classes:
+                            proba_y[col] = 0
+                    proba_y = proba_y[task.class_labels]
             else:
                 raise ValueError("The task has no class labels")
 
+            if not isinstance(proba_y, pd.DataFrame):
+                proba_y = pd.DataFrame(proba_y, columns=task.class_labels)
```

Comment:
Could you please add an elif here that checks that the columns of the dataframe are equal to the class labels?

Reply:
Sure, certainly.

Reply:
Added [here](https://github.com/openml/openml-python/blob/c06eb0da24498c08c50edcf579075c7e500a5301/openml/extensions/sklearn/extension.py#L1697).

Reply:
Not sure if it's necessary, but we could also check that the order stays the same